### PR TITLE
Added public/private encrypt/decrypt functions to RSA

### PR DESCRIPTION
**Pull request by aoprisan**
This is about issue #460 . I have ported partially the code in the commit to PKey, including tests. I added the following functions https://github.com/thommay/rust-openssl/blob/11e3b1b56317ca1e24cfc1c0a3805123fa73bfb8/openssl/src/crypto/pkey.rs#L393, though I am not sure all are needed. 
I also added two macros to handle OpenSSL methods which return size values (-1 is the error code).

### CODE REVIEW

**Comment by bbatha** [comment]
Instead of the new macro you can use `try_ssl_if!(err_code == -1)`

File: openssl/src/crypto/rsa.rs [hunk 1]
```diff
@@ -162,6 +180,125 @@ impl RSA {
         }
     }
 
+    pub fn max_data(&self) -> Option<u32> {
+        // 41 comes from RSA_public_encrypt(3) for OAEP
+        self.size().map(|len| len - 41)
```
Comment (sfackler) [inline openssl/src/crypto/rsa.rs]: This was copied from the old impl, right? There's a (now closed) issue talking about how this is wrong for non-OAEP padding, so we shouldn't have the method unless we can implement it correctly.

File: openssl/src/crypto/rsa.rs [hunk 2]
```diff
@@ -11,6 +11,24 @@ use HashTypeInternals;
...
+/// Type of encryption padding to use.
+#[derive(Copy, Clone)]
+pub enum EncryptionPadding {
+    NoPadding,
+    OAEP,
+    PKCS1v15
+}
+
+impl EncryptionPadding {
+    pub fn openssl_padding_code(&self) -> c_int {
```
Comment (sfackler) [inline openssl/src/crypto/rsa.rs]: Can we make this private?

File: openssl/src/crypto/rsa.rs [hunk 3]
```diff
@@ -162,6 +180,125 @@ impl RSA {
...
+
+           Ok(to)
+        }
+    }
+
+    /**
+      * Encrypts data with the public key, using OAEP padding, returning the encrypted data. The
+      * supplied data must not be larger than max_data().
+      */
+     pub fn encrypt(&self, s: &[u8]) -> Result<Vec<u8>, ErrorStack> { self.public_encrypt_with_padding(s, EncryptionPadding::OAEP) }
```
Comment (sfackler) [inline openssl/src/crypto/rsa.rs]: I would just not have this and force people to specify their padding personally.

**Comment by sfackler** [comment]
Thanks! Couple of little tweaks and this should be good to go.

File: openssl/src/crypto/rsa.rs (hunk 2 above)
Comment (aoprisan) [inline openssl/src/crypto/rsa.rs]: Sure, will do.

File: openssl/src/crypto/rsa.rs (hunk 1 above)
Comment (aoprisan) [inline openssl/src/crypto/rsa.rs]: Yep, I will remove it for now.

Comment (aoprisan) [inline openssl/src/crypto/rsa.rs]: Fixing it would involve passing the encoding and taking that into account ?

**Comment by aoprisan** [comment]
`try_ssl_if` returns Unit and in my case I need the returned value. Should I leave `lift_ssl_returns_size` or refactor `try_ssl_if` ?

File: openssl/src/crypto/rsa.rs (hunk 1 above)
Comment (sfackler) [inline openssl/src/crypto/rsa.rs]: Yep, I believe so.

**Comment by sfackler** [comment]
I've been meaning to refactor the error conversion macros, don't worry about that for now.

**Comment by sfackler** [comment]
Thanks!

**Pull request merged**